nodebuilder/blobstream/service.go

```diff
@@ -0,0 +1,534 @@
+	commitmentProof := CommitmentProof{
+		SubtreeRoots:      subtreeRoots,
+		SubtreeRootProofs: nmtProofs,
+		NamespaceID:       namespace.ID(),
+		RowProof:          sharesProof.RowProof,
+		NamespaceVersion:  namespace.Version(),
+	}
+
+	return &ResultCommitmentProof{CommitmentProof: commitmentProof}, nil
+}
+
+// computeSubtreeRoots takes a set of shares and ranges and returns the corresponding subtree roots.
+// the offset is the number of shares that are before the subtree roots we're calculating.
+func computeSubtreeRoots(shares []share.Share, ranges []nmt.LeafRange, offset int) ([][]byte, error) {
+	if len(shares) == 0 {
+		return nil, fmt.Errorf("cannot compute subtree roots for an empty shares list")
+	}
+	if len(ranges) == 0 {
+		return nil, fmt.Errorf("cannot compute subtree roots for an empty ranges list")
+	}
+	if offset < 0 {
+		return nil, fmt.Errorf("the offset %d cannot be stricly negative", offset)
+	}
+
+	// create a tree containing the shares to generate their subtree roots
+	tree := nmt.New(appconsts.NewBaseHashFunc(), nmt.IgnoreMaxNamespace(true), nmt.NamespaceIDSize(share.NamespaceSize))
+	for _, sh := range shares {
+		var leafData []byte
+		leafData = append(append(leafData, share.GetNamespace(sh)...), sh...)
+		err := tree.Push(leafData)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// generate the subtree roots
+	var subtreeRoots [][]byte
+	for _, rg := range ranges {
+		root, err := tree.ComputeSubtreeRoot(rg.Start-offset, rg.End-offset)
+		if err != nil {
+			return nil, err
+		}
+		subtreeRoots = append(subtreeRoots, root)
+	}
+	return subtreeRoots, nil
+}
+
+func uint64ToInt(number uint64) (int, error) {
+	if number >= math.MaxInt {
+		return 0, fmt.Errorf("number %d is higher than max int %d", number, math.MaxInt)
+	}
+	return int(number), nil
+}
+
+// ProveSubtreeRootToCommitment generates a subtree root to share commitment inclusion proof.
+// Note: this method is not part of the API. It will not be served by any endpoint, however,
+// it can be called directly programmatically.
+func ProveSubtreeRootToCommitment(subtreeRoots [][]byte, subtreeRootIndex uint64) (*ResultSubtreeRootToCommitmentProof, error) {
+	_, proofs := merkle.ProofsFromByteSlices(subtreeRoots)
+	return &ResultSubtreeRootToCommitmentProof{
+		SubtreeRootToCommitmentProof: SubtreeRootToCommitmentProof{
+			Proof: *proofs[subtreeRootIndex],
+		},
+	}, nil
+}
+
+// ProveShareToSubtreeRoot generates a share to subtree root inclusion proof
+// Note: this method is not part of the API. It will not be served by any endpoint, however,
+// it can be called directly programmatically.
+func ProveShareToSubtreeRoot(shares [][]byte, shareIndex uint64) (*ResultShareToSubtreeRootProof, error) {
+	_, proofs := merkle.ProofsFromByteSlices(shares)
+	return &ResultShareToSubtreeRootProof{
+		ShareToSubtreeRootProof: ShareToSubtreeRootProof{
+			Proof: *proofs[shareIndex],
+		},
+	}, nil
```

Comment:
[note to reviewers]
these two methods are not used in the API but I see them as useful for people using the API. should we keep them here? move them somewhere special? create a `utils.go` or `helpers.go` and put them there?